Add pass to remove unused parameters in to_edge

GregoryComer opened the pull request:
Summary:
Currently, ExecuTorch will serialize any parameters in the exported program, regardless of whether they are actually used. Exporting with strict=True will remove unused parameters, but strict=False will not. Export recently switched to non-strict as the default behavior.

This causes PTE bloat when doing pt2e quantization (unquantized weights are left in the graph) or sometimes when exporting multiple methods (encode and decoder, for example).

This PR adds a new pass (`remove_unused_parameters_pass`) to strip unused parameters from the `ExportedProgram`. It is run as part of `to_executorch`. Parameters are considered unused if there are no uses of the placeholder node. Parameters are removed by stripping them from the state_dict, input specs, and graph.

As a question for reviewers, should we run this pass earlier, as part of to_edge? My rationale for running as part of to_executorch was that it could theoretically clean up anything else left by partitioning and lowering, but I'm not aware of any concrete use cases for this.

Differential Revision: D73654202




cc @JacobSzwejbka @angelayi

JacobSzwejbka commented:
> It is run as part of to_executorch

Can we run it in to_edge instead? 

We should remove unused things as soon as possible imo.

JacobSzwejbka commented on exir/passes/remove_unused_parameters_pass.py:
```diff
@@ -0,0 +1,54 @@
+) -> bool:
+    placeholder_node = placeholder_nodes.get(parameter)
+    if placeholder_node is None:
+        # Shouldn't happen, but in this case, leave the parameter to be safe.
```
throw an exception?

GregoryComer committed:
Add pass to remove unused parameters in to_executorch (#10484)

Summary:
Currently, ExecuTorch will serialize any parameters in the exported program, regardless of whether they are actually used. Exporting with strict=True will remove unused parameters, but strict=False will not. Export recently switched to non-strict as the default behavior.

This causes PTE bloat when doing pt2e quantization (unquantized weights are left in the graph) or sometimes when exporting multiple methods (encode and decoder, for example).

This PR adds a new pass (`remove_unused_parameters_pass`) to strip unused parameters from the `ExportedProgram`. It is run as part of `to_executorch`. Parameters are considered unused if there are no uses of the placeholder node. Parameters are removed by stripping them from the state_dict, input specs, and graph.

As a question for reviewers, should we run this pass earlier, as part of to_edge? My rationale for running as part of to_executorch was that it could theoretically clean up anything else left by partitioning and lowering, but I'm not aware of any concrete use cases for this.

cc JacobSzwejbka angelayi

Pull Request resolved: https://github.com/pytorch/executorch/pull/10484

Reviewed By: digantdesai, JacobSzwejbka

Differential Revision: D73654202

Pulled By: GregoryComer
